packages/server/test/unit/ensure_url_spec.ts

```diff
@@ -0,0 +1,61 @@
+require('../spec_helper')
+
+import { connect, agent } from '@packages/networking'
+import { ensureUrl } from '../../lib/util/ensure-url'
+
+describe('lib/util/ensure-url', function() {
+  context("#ensureUrl", function() {
+    it("resolves if a URL connects", function() {
+      const stub = sinon.stub(connect, 'getAddress').withArgs(80, 'foo.bar.invalid').resolves()
+
+      return ensureUrl('http://foo.bar.invalid')
+      .then(() => {
+        expect(stub).to.be.calledOnce
+      })
+    })
+
+    it("rejects if a URL doesn't connect", function() {
+      const stub = sinon.stub(connect, 'getAddress').withArgs(80, 'foo.bar.invalid').rejects()
+
+      return ensureUrl('http://foo.bar.invalid')
+      .then(() => {
+        const err: any = new Error('should not reach this')
+        err.fromTest = true
+      })
+      .catch((e) => {
+        if (e.fromTest) {
+          throw e
+        }
+        expect(stub).to.be.calledOnce
+      })
+    })
+  })
+
+  context('with a proxy', function() {
+    beforeEach(function() {
+      this.oldEnv = Object.assign({}, process.env)
+    })
+
+    afterEach(function() {
+      process.env = this.oldEnv
+    })
+
+    it('calls into the agent to check availability', function() {
+      process.env.HTTP_PROXY = process.env.HTTPS_PROXY = 'http://localhost:12345'
+      process.env.NO_PROXY = ''
+
+      const spy = sinon.stub(agent, 'addRequest').throws()
+      nock.enableNetConnect()
+
+      return ensureUrl('http://foo.bar.invalid')
+      .then(() => {
+        throw new Error('should not succeed')
+      })
+      .catch((e) => {
+        expect(spy).to.be.calledOnce
```

Comment:
this is a `stub` not a spy. with sinon you can actually create a separate thing which is called a `spy` - so you'd always name your thing based on which one you created.

alternatively you don't have to assign it to anything if you don't want..

```js
const stub = sinon.stub(a, 'b')
const spy = sinon.spy(a, 'b')

// or just...
sinon.stub(a, 'b')

a.b()

expect(a.b).to.be.calledOnce
```

Reply:
oops, that was originally a `spy` call :sweat_smile: good catch. i'll update this

Reply:
I'm keeping a reference to the calls because this way, stub has a type of `Sinon.SinonStub` instead of `agent.addRequest`'s signature and I can use `getCall` and stuff on it without TypeScript getting mad.